Show of `Array{Tuple{Int64,Array{Float64,1}},1}` broken on master

```
julia> versioninfo()
Julia Version 0.7.0-DEV.2832
Commit feec0ecf1c* (2017-12-12 03:49 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-7660U CPU @ 2.50GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-4.0.0 (ORCJIT, skylake)
Environment:
```

```
julia> fill((1, zeros(3)), 10)
10-element Array{Tuple{Int64,Array{Float64,1}},1}:
Error showing value of type Array{Tuple{Int64,Array{Float64,1}},1}:
ERROR: AssertionError: Array{Float64,1} is not a subtype of Tuple{Int64,Array{Float64,1}}
Stacktrace:
 [1] typeinfo_prefix(::IOContext{Base.GenericIOBuffer{Array{UInt8,1}}}, ::Array{Float64,1}) at ./arrayshow.jl:475
 [2] show_vector(::IOContext{Base.GenericIOBuffer{Array{UInt8,1}}}, ::Array{Float64,1}, ::Char, ::Char) at ./arrayshow.jl:428 (repeats 2 times)
 [3] show(::IOContext{Base.GenericIOBuffer{Array{UInt8,1}}}, ::Array{Float64,1}) at ./arrayshow.jl:442
 [4] show_delim_array(::IOContext{Base.GenericIOBuffer{Array{UInt8,1}}}, ::Tuple{Int64,Array{Float64,1}}, ::Char, ::Char, ::Char, ::Bool, ::Int64, ::Int64) at ./show.jl:501
 [5] show_delim_array at ./show.jl:489 [inlined]
 [6] show(::IOContext{Base.GenericIOBuffer{Array{UInt8,1}}}, ::Tuple{Int64,Array{Float64,1}}) at ./show.jl:517
 [7] #sprint#283(::IOContext{Base.Terminals.TTYTerminal}, ::Function, ::Int64, ::Function, ::Tuple{Int64,Array{Float64,1}}, ::Vararg{Tuple{Int64,Array{Float64,1}},N} where N) at ./strings/io.jl:78
 [8] #sprint at ./<missing>:0 [inlined]
 [9] alignment(::IOContext{Base.Terminals.TTYTerminal}, ::Tuple{Int64,Array{Float64,1}}) at ./show.jl:1584
 [10] alignment(::IOContext{Base.Terminals.TTYTerminal}, ::Array{Tuple{Int64,Array{Float64,1}},1}, ::Base.OneTo{Int64}, ::Base.OneTo{Int64}, ::Int64, ::Int64, ::Int64) at ./arrayshow.jl:68
 [11] print_matrix(::IOContext{Base.Terminals.TTYTerminal}, ::Array{Tuple{Int64,Array{Float64,1}},1}, ::String, ::String, ::String, ::String, ::String, ::String, ::Int64, ::Int64) at ./arrayshow.jl:187
 [12] print_matrix at ./arrayshow.jl:160 [inlined]
 [13] print_array at ./arrayshow.jl:309 [inlined]
 [14] _display(::IOContext{Base.Terminals.TTYTerminal}, ::Array{Tuple{Int64,Array{Float64,1}},1}) at ./arrayshow.jl:340
 [15] show(::IOContext{Base.Terminals.TTYTerminal}, ::MIME{Symbol("text/plain")}, ::Array{Tuple{Int64,Array{Float64,1}},1}) at ./replutil.jl:139
 [16] display(::Base.REPL.REPLDisplay{Base.REPL.LineEditREPL}, ::MIME{Symbol("text/plain")}, ::Array{Tuple{Int64,Array{Float64,1}},1}) at ./repl/REPL.jl:126
 [17] display(::Base.REPL.REPLDisplay{Base.REPL.LineEditREPL}, ::Array{Tuple{Int64,Array{Float64,1}},1}) at ./repl/REPL.jl:129
 [18] display(::Array{Tuple{Int64,Array{Float64,1}},1}) at ./multimedia.jl:291
 [19] (::getfield(Base, Symbol("#inner#4")){NamedTuple{(),Tuple{}},typeof(display),Tuple{Array{Tuple{Int64,Array{Float64,1}},1}}})() at ./essentials.jl:665
 [20] #invokelatest#3 at ./essentials.jl:666 [inlined]
 [21] invokelatest at ./essentials.jl:665 [inlined]
 [22] print_response(::Base.Terminals.TTYTerminal, ::Any, ::Void, ::Bool, ::Bool, ::Void) at ./repl/REPL.jl:147
 [23] print_response(::Base.REPL.LineEditREPL, ::Any, ::Void, ::Bool, ::Bool) at ./repl/REPL.jl:133
 [24] (::getfield(Base.REPL, Symbol("#do_respond#17")){Bool,getfield(Base.REPL, Symbol("##27#37")){Base.REPL.LineEditREPL,Base.REPL.REPLHistoryProvider},Base.REPL.LineEditREPL,Base.LineEdit.Prompt})(::Base.LineEdit.MIState, ::Base.GenericIOBuffer{Array{UInt8,1}}, ::Bool) at ./repl/REPL.jl:704
 [25] top-level scope
```

--- a/base/show.jl
+++ b/base/show.jl
@@ -491,14 +491,18 @@ function show_delim_array(io::IO, itr, op, delim, cl, delim_one, i1=1, n=typemax
         recur_io = IOContext(io, :SHOWN_SET => itr)
         state = start(itr)
         first = true
+        i0 = i1-1
         while i1 > 1 && !done(itr, state)
             _, state = next(itr, state)
             i1 -= 1
         end
         if !done(itr, state)
+            typeinfo = get(io, :typeinfo, Any)
             while true
                 x, state = next(itr, state)
-                show(recur_io, x)
+                show(IOContext(recur_io, :typeinfo =>
+                               typeinfo <: Tuple ? typeinfo.parameters[i1+i0] : typeinfo),
+                     x)
                 i1 += 1
                 if done(itr, state) || i1 > n
                     delim_one && first && print(io, delim)

--- a/test/show.jl
+++ b/test/show.jl
@@ -1035,6 +1035,9 @@ end
     @test showstr([[Float16(1)]]) == "Array{Float16,1}[[1.0]]"
     @test replstr(Real[Float16(1)]) == "1-element Array{Real,1}:\n Float16(1.0)"
     @test replstr(Array{Real}[Real[1]]) == "1-element Array{Array{Real,N} where N,1}:\n [1]"
+    # printing tuples (Issue #25042)
+    @test replstr(fill((Int64(1), zeros(Float16, 3)), 1)) ==
+                 "1-element Array{Tuple{Int64,Array{Float16,1}},1}:\n (1, [0.0, 0.0, 0.0])"
     @testset "nested Any eltype" begin
         x = Any[Any[Any[1]]]
         # The element of x (i.e. x[1]) has an eltype which can't be deduced

